Invalidate implicit resolution of JsonImplicits.formatMap

jeffmay opened the pull request:
@vsrally @rallyhealth/engineers 

This method is both `implicit` and unsound. It is a landmine, and we need to remove it ASAP.

This first change will invalidate the implicit resolution of formatMap without breaking binary compatibility.

Future changes will be to add support for an improved version of this implicit, remove support for Play 2.3, and then add support for Play 2.6 and 2.7.0-M1.

jeffmay commented on play25-json-ops/src/main/scala/play/api/libs/json/ops/JsonImplicits.scala:
```diff
@@ -20,7 +20,8 @@ trait JsonImplicits extends ImplicitTupleFormats with JsValueImplicits {
    */
-  implicit def formatMap[K, V: Format](
+  @deprecated("This method is unsafe and will be removed (possibly replaced by a more sound implementation).", "2.0.0")
+  def formatMap[K, V: Format](
```
I think this will end up being a breaking change anyway. We might want to just release 3.0.0 with the correct fix.

jeffmay replied:
If I'm gonna go with a 3.0.0 fix, then it probably makes sense to pull out a common `playX-json-ops-common` project with cross-compatible play-json code. Also, I should add support for Play 2.6 / Scala 2.12 and Play 2.7.0-M1 / Scala 2.13.0-M4

slustbader commented:
Why is it unsafe?

slustbader commented:
I would vote for fixing it properly or removing it entirely if it's really that bad.

vsrally reviewed (approved):
Maybe add a note on the README about it too

jeffmay commented:
My current plan is to release a 3.0.0 version with a fixed version of this method. I'm also going to support Play 2.6.10 and 2.7.0-M1 in the 3.0.0 version.

@slustbader the issue was that this function accepts implicit conversions to / from string, and even implements an unsound default parameter that uses `.toString` (YIKES!) The solution is to use a special `ReadsKey` / `WritesKey` traits that convert to `String`. This will prevent this implicit from getting picked up and accidentally serializing json objects unsoundly.

jeffmay committed:
Invalidate implicit resolution for JsonImplicits.formatMap

- This removes an unsafe implicit landmine that would call .toString
  as a default implicit argument
- These changes are still binary compatible
- Add unit tests for ambiguous implicit fix

jeffmay commented:
Ok, I have a working branch with Play 2.7 / Scala 2.13.0-M4 support, I'm going to start pushing the commits through to master.

This first change will invalidate the implicit resolution of formatMap without breaking binary compatibility.

The remaining changes will be to add support for an improved version of this implicit, remove support for Play 2.3, and then add support for Play 2.6 and 2.7.0-M1.

jeffmay committed:
Update build to use codecov.io instead of coveralls

jeffmay committed:
Update README.md to use new badges
